opencart 1.5.6.4/upload/catalog/model/smailyforopencart/helper.php

```diff
@@ -0,0 +1,112 @@
+        // Credentials.
+        $username = $settings['username'];
+        $password = $settings['password'];
+        $subdomain = $settings['subdomain'];
+        // Url.
+        $apiUrl = 'https://' . $subdomain . '.sendsmaily.net/api/' . $endpoint . '.php';
+        // cURL call.
+        $ch = curl_init();
+        curl_setopt($ch, CURLOPT_URL, $apiUrl);
+        curl_setopt($ch, CURLOPT_HEADER, 0);
+        curl_setopt($ch, CURLOPT_RETURNTRANSFER, 1);
+        curl_setopt($ch, CURLOPT_HTTPAUTH, CURLAUTH_BASIC);
+        curl_setopt($ch, CURLOPT_USERPWD, "$username:$password");
+
+        // API call with GET request.
+        if ($method === 'GET') {
+            if (!empty($data)) {
+                curl_setopt($ch, CURLOPT_URL, $apiUrl . '?' . http_build_query($data));
+            }
+            $api_call = json_decode(curl_exec($ch), true);
+            // Response code from Smaily API.
+            $http_code = curl_getinfo($ch, CURLINFO_HTTP_CODE);
+        // If Method POST.
+        } elseif ($method === 'POST') {
+            curl_setopt($ch, CURLOPT_POST, 1);
+            curl_setopt($ch, CURLOPT_POSTFIELDS, http_build_query($data));
+            $api_call = json_decode(curl_exec($ch), true);
+            // Response code from Smaily API.
+            $http_code = curl_getinfo($ch, CURLINFO_HTTP_CODE);
+            // Validate response.
+            if (!array_key_exists('code', $api_call)) {
+                $this->log->write('Something went wrong with Smaily request!');
+            }
+            if (isset($api_call['code']) && (int) $api_call['code'] !== 101) {
+                $this->log->write($api_call['message']);
+            }
+        }
+        // Return response if success.
+        if ($http_code === 200) {
+            // POST.
+            if ($method === 'POST') {
+                if (array_key_exists('code', $api_call) && (int) $api_call['code'] === 101) {
+                    $response = $api_call;
+                }
+            // GET.
+            } else {
+                $response = $api_call;
+            }
+        } else {
+            // Log error.
+            $this->log->write('Error in smaily api call with code: ' . $http_code);
+        }
+        // Response from API call.
+        return $response;
+    }
+
+    /**
+     * Get all subscribed customers.
+     *
+     * @return array $customers All subscribed customers in array.
+     */
+    public function getSubscribedCustomers($offset) {
+        $query = $this->db->query("SELECT * FROM " . DB_PREFIX . "customer
+            WHERE (`customer_id` > " . (int)$offset . " AND `newsletter` = '1') LIMIT 2500");
+        return $query->rows;
+    }
+
+    /**
+     * Sets newsletter status to 0 in customer table.
+     *
+     * @param array $emails Emails to unsubscribe.
+     *
+     * @return void
+     */
+    public function unsubscribeCustomers($emails) {
```

Comment:
There are issues with this implementation:
a) this does not resolve the issue with updating subscribers one-by-one;
b) `$email` is not escaped which will make this module vulnerable to SQL injection attacks;
c) `customer_id <> 0` does have any effect on the query - `customer_id` cannot be anything other than a unique number, and there is no index on `customer_id` and `email` columns to speed up the query.

Method should be implemented as:
```php
$binds = array();
foreach ($emails as $email) {
    $binds[] = $this->db->escape($email);
}

$this->db->query(
    "UPDATE " . DB_PREFIX . "customer SET newsletter = '0' ".
    "WHERE `email` IN ('" . implode("','", $binds) . "')");
```

If there should be an issue with query being too long, you can wrap method's implementation with:
```php
$chunks = array_chunk($emails, 500);
foreach ($chunks as $chunk) {
    $binds = array()
    // ...

    $this->db->query("UPDATE ...");
}
```